enhancements/machine-api/gcp-spot-instances.md

```diff
@@ -0,0 +1,172 @@
+
+- Graceful termination of nodes should be provided by observing termination notices
+
+### Implementation Details
+
+#### Termination handler
+
+The termination handler is already implemented and it checks if the instance has the label `interruptible-instance`. If it does, the handler will make sure the instance will have a graceful termination.
+
+Both GCP instance types, legacy preemptible and spot, have the same termination process based on the [Spot instances termination docs](https://cloud.google.com/compute/docs/instances/spot#preemption-process) and [Legacy preemptible instances termination docs](https://cloud.google.com/compute/docs/instances/preemptible#preemption-process). This means no changes are required in the termination handler, besides the instance having the label `interruptible-instance`.
+
+
+#### Implementation Specifics
+
+
+###### Launching GCP spot instances
+
+To launch a spot instance on GCP, the `ProvisioningModel` field must be set:
+
+```go
+&compute.Instance{
+  ...
+  Scheduling: &compute.Scheduling{
+    ...
+    ProvisioningModel: r.providerSpec.ProvisioningModel,
+  },
+}
+```
+
+Therefore, to make the choice up to the user, this field should be added to the `GCPMachineProviderSpec`:
+
+```go
+type GCPMachineProviderSpec struct {
+  ...
+  ProvisioningModel GCPProvisioningModelType `json:”provisioningModel”`
+}
+```
+
+Once the instance is launched, the Machine will be labelled as an `interruptible-instance`
+if the instance `Scheduling.ProvisioningModel` field is set to `SPOT`.
+
+
+### Risks and Mitigations
+
+
+## Design Details
+
+### Test Plan
+
+### Graduation Criteria
+
+#### Examples
+
+##### Dev Preview -> Tech Preview
+
+##### Tech Preview -> GA
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Implementation History
+
+## Drawbacks
+
+
+## Related Research
+
+
+### Limitations of Preemptible
+
+#### Legacy preemptible instance limitation
+
+Legacy preemptible instances will, if not already, be terminated after 24 hours.
+This means that the instances will be cycled regularly and as such, good handling of shutdown events should be implemented.
+
+#### Shutdown warning
+GCP gives a 30 second warning for termination of Preemptible instances.
+This signal comes via an ACPI G2 soft-off signal to the machine, which, could be intercepted to start a graceful termination of pods on the machine.
+There are [existing projects](https://github.com/GoogleCloudPlatform/k8s-node-termination-handler) that already do this.
+
+Alternatively, GCP provides a [metadata service](https://cloud.google.com/compute/docs/instances/create-start-preemptible-instance#detecting_if_an_instance_was_preempted)
+that is accessible from the instances themselves that allows the instance to determine whether or not it has been preempted.
+This is similar to what is provided by AWS and Azure and should be used to allow the termination handler implementation
+to be consistent across the three providers.
```

Comment:
This is new since we implemented the termination handler originally, if we think we can move the termination handler to use this, then we should do that to reduce the drift between the different implementations (it's also easier to mock the termination endpoint as opposed to the ACPI soft off)

Reply:
is it? as it's `related research`, I basically copied this from the original proposal. You can check here https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/spot-instances.md#limitations-of-preemptible
am I missing something here?

Reply:
Ok, good point, I have no idea why the GCP one was implemented differently now then 🤔 You can ignore this comment but we probably do want to look at re-aligning them at some point